### PR TITLE
Fix for NUTCH-2450 by Kenneth McFarland

### PR DESCRIPTION
**Pull request by kpm1985**

### CODE REVIEW

File: src/java/org/apache/nutch/parse/ParseOutputFormat.java [hunk 1]
```diff
@@ -379,15 +378,16 @@ public static String filterNormalize(String fromUrl, String toUrl,
...
-          //FIXME: toDomain will never be null, correct?
           if (toDomain == null || toDomain.equals(origin)) {
             return null; // skip it
           }
         } else {
           String toHost = targetURL.getHost().toLowerCase();
-          //FIXME: toDomain will never be null, correct?
           if (toHost == null || toHost.equals(origin)) {
-            return null; // skip it
+            if (exemptionFilters == null // check if it is exempted?
```
Comment (sebastian-nagel) [inline src/java/org/apache/nutch/parse/ParseOutputFormat.java]: The [exemption filter|https://builds.apache.org/job/nutch-trunk/javadoc/org/apache/nutch/urlfilter/ignoreexempt/package-summary.html] is to define exemption for otherwise skipped external links. Should not be in the branch which handles internal links, otherwise we would have to redefine the exemption filter interface.

File: src/java/org/apache/nutch/parse/ParseOutputFormat.java [hunk 2]
```diff
@@ -362,7 +362,6 @@ public static String filterNormalize(String fromUrl, String toUrl,
       if (ignoreExternalLinks) {
         if ("bydomain".equalsIgnoreCase(ignoreExternalLinksMode)) {
           String toDomain = URLUtil.getDomainName(targetURL).toLowerCase();
-          //FIXME: toDomain will never be null, correct?
```
Comment (sebastian-nagel) [inline src/java/org/apache/nutch/parse/ParseOutputFormat.java]: Please check whether URLUtil.getDomainName(URL url) may return null and add this information to the javadoc of the function. The FIXME should then be resolved accordingly.
If you're uncertain about the behavior, you could a test to TestURLUtil.testGetDomainDame() to test the actual behavior, e.g.,
```
    // test URL without host/authority
    url = new URL("file:/path/index.html");
    Assert.assertNull(URLUtil.getDomainName(url));
    Assert.assertNotNull(URLUtil.getDomainName(url));
    Assert.assertEquals("", URLUtil.getDomainName(url));
```
Of course, not all three assertions may succeed. Please also add such a test to your PR, or even more for edge cases of URLs without host or domain.

File: src/java/org/apache/nutch/parse/ParseOutputFormat.java (hunk 1 above)
Comment (kpm1985) [reply src/java/org/apache/nutch/parse/ParseOutputFormat.java]: Thank you for this review @sebastian-nagel , I will be attacking this over this weekend. I appreciate the clarification in understanding. This project has absolutely fantastic reviewers, thank you again!

**Comment by lewismc** [comment]
@kpm1985 ping, trying to close some issues here.

**Comment by kpm1985** [comment]
@lewismc Apologies. Where would you like the unit test to be located?

**Comment by lewismc** [comment]
[TestURLUtil](https://github.com/apache/nutch/blob/master/src/test/org/apache/nutch/util/TestURLUtil.java) testGetDomainDame() or additional tests for that test class.

**Comment by kpm1985** [comment]
I closed the issue. The code is still labeled "FIXME", I don't want to mess up closing issues @lewismc  and since Eclipse has been being such a frustration issue I am just going to take a break from the time + frustration issues and come back. 

I hope that is ok, I have taken notes and will come back with a preemptive strike next time. I'm just not good with pressure right now. Thank you for giving me the direction, the next time I PR it will have more thought because of this.
